### PR TITLE
Feature adding pagination to sigh profile downloading

### PR DESCRIPTION
**Pull request by LizCira**
This is the functionality discussed with @TKBurner in this issue: https://github.com/fastlane/fastlane/issues/4605#issuecomment-219127594

Problem
Currently, running "`sigh download_all`" consistently times out when used with Apple Developer accounts containing 1k+ profiles. 

The stacktrace for the error is: 

```
usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/protocol.rb:158:in `rescue in rbuf_fill': [!] Net::ReadTimeout (Faraday::TimeoutError)
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/protocol.rb:152:in `rbuf_fill'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/protocol.rb:134:in `readuntil'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/protocol.rb:144:in `readline'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/http/response.rb:39:in `read_status_line'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/http/response.rb:28:in `read_new'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/http.rb:1408:in `block in transport_request'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/http.rb:1405:in `catch'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/http.rb:1405:in `transport_request'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/http.rb:1378:in `request'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/http.rb:1371:in `block in request'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/http.rb:853:in `start'
    from /usr/local/rvm/rubies/ruby-2.1.2/lib/ruby/2.1.0/net/http.rb:1369:in `request'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:82:in `perform_request'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:40:in `block in call'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:87:in `with_net_http_connection'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:32:in `call'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday-cookie_jar-0.0.6/lib/faraday/cookie_jar.rb:22:in `call'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response_middleware.rb:30:in `call'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response_middleware.rb:30:in `call'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response_middleware.rb:30:in `call'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday-0.9.2/lib/faraday/rack_builder.rb:139:in `build_response'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday-0.9.2/lib/faraday/connection.rb:377:in `run_request'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/faraday-0.9.2/lib/faraday/connection.rb:177:in `post'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/spaceship-0.26.1/lib/spaceship/client.rb:404:in `block in send_request'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/spaceship-0.26.1/lib/spaceship/client.rb:322:in `with_retry'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/spaceship-0.26.1/lib/spaceship/client.rb:403:in `send_request'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/spaceship-0.26.1/lib/spaceship/client.rb:379:in `request'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/spaceship-0.26.1/lib/spaceship/portal/portal_client.rb:317:in `provisioning_profiles'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/spaceship-0.26.1/lib/spaceship/portal/provisioning_profile.rb:246:in `all'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/sigh-1.6.1/lib/sigh/download_all.rb:10:in `download_all'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/sigh-1.6.1/lib/sigh/manager.rb:34:in `download_all'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/sigh-1.6.1/bin/sigh:45:in `block (2 levels) in run'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/commander-4.3.5/lib/commander/command.rb:178:in `call'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/commander-4.3.5/lib/commander/command.rb:178:in `call'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/commander-4.3.5/lib/commander/command.rb:153:in `run'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/commander-4.3.5/lib/commander/runner.rb:428:in `run_active_command'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/fastlane_core-0.41.2/lib/fastlane_core/ui/fastlane_runner.rb:23:in `run!'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/commander-4.3.5/lib/commander/delegates.rb:15:in `run!'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/sigh-1.6.1/bin/sigh:93:in `run'
    from /usr/local/rvm/gems/ruby-2.1.2/gems/sigh-1.6.1/bin/sigh:111:in `<top (required)>'
    from /usr/local/rvm/gems/ruby-2.1.2/bin/sigh:23:in `load'
    from /usr/local/rvm/gems/ruby-2.1.2/bin/sigh:23:in `<main>'
    from /usr/local/rvm/gems/ruby-2.1.2/bin/ruby_executable_hooks:15:in `eval'
    from /usr/local/rvm/gems/ruby-2.1.2/bin/ruby_executable_hooks:15:in `<main>'
```

Solution
This PR adds the already existing "page" pagination wrapper to Spaceship's portal_client.rb. Pagination is what was causing the response from the Apple dev portal to fail.

Questions for Review
The page count default for the @page variable is originally set to 500 items per page, and I've changed it to 30 for downloading profiles. My tests showed that the Apple dev portal calls began timing out around 300 profiles per page but that performance slowed down around 100, fyi. I'm not sure if any other code depends on this variable being set to 500. If so, this could raise problems in a multi-thread scenario. 

I've also added a feature to Sigh that's an optional flag allowing users to keep original profile names when downloading profiles. This is useful if you're using download_all to sync profiles across several machines. If there's objection to this flag I'll remove it for the pull request.  

Excited to contribute, thanks guys!

### CODE REVIEW

**Comment by KrauseFx** [comment]
Hey @LizCira, thanks for the PR, but quality wise there are a lot of issues that need to be solved first
1) Code Style issues, please run `rubocop -a` and make sure everything is green
2) There are no new tests added
3) `Uses the original Apple Developer portal profile names, instead of Fastlane's generated names. For use with 'sigh download_all.'`. Does this really only work with `download_all`?

**Comment by LizCira** [comment]
Thanks for the response, sorry it's taken a few days to get back to this. @asfalcone (or @KrauseFx who must be really busy) is it expected behavior for the circleci tests to pass here? 

For Felix's points. 

1. Fixed style violations with rubocop and tested behavior still works as expected.

2. I'm not too experienced with writing stubbing rspec tests. I'll try to look at this tomorrow. If you can point me towards any tutorials or helpful resources, much obliged!  

3.
>  "Uses the original Apple Developer portal profile names, instead of Fastlane's generated names. For use with 'sigh download_all.'. Does this really only work with download_all?"

There's no other "sigh" commands to download profiles, are there? I don't see see any command to download an individual profile, so I'm not sure what else it could be used with.

**Comment by KrauseFx** [comment]
Hey @LizCira, thanks for all you work so far and sorry for the late reply. Looking at the Files Changed, there seems to be something wrong with this PR. Can you make sure to rebase from master and only show the actual change from the `master` branch?

**Comment by sfransen** [comment]
When is this Pull request going to be pushed. I have tested the code and it works great on large accounts. I would make one change but I will make that pull request after this is pushed.

**Comment by ohwutup** [comment]
Hey @LizCira - just wanted to check in to see if you had a chance to rebase on master and then push up the branch again?  🚀 

Hey @sfransen - we are just waiting for the build to succeed before we can move forward. Would you mind suggesting your change here so that we can discuss it and decide whether or not we should try to get it in all at once! The more opinions the merrier! 😄

**Comment by sfransen** [comment]
The 30 needs to be changed to 10 as my license just started failing because the pagination is too large now.

  def provisioning_profiles(mac: false)
      @page_size = 10  # was 30
      paging do |page_number|
        puts "Pagination in progress, page number: '#{page_number}'..."
        req = request(:post) do |r|
          r.url "https://developerservices2.apple.com/services/#{PROTOCOL_VERSION}/#{platform_slug(mac)}/listProvisioningProfiles.action"
          r.params = {
            teamId: team_id,
            includeInactiveProfiles: true,
            onlyCountLists: true,
            pageSize: page_size,
            pageNumber: page_number,
            sort: "name=dasc"
          }
        end

        parse_response(req, 'provisioningProfiles')
      end
    end

**Comment by LizCira** [comment]
Hi all, apologies, I've been traveling--I will clean up the PR and resubmit this tomorrow. I'll update this comment. 

@sfransen what do you mean your license started failing? Can you give more details? It was still working on our build servers this morning. Does it have something to do with WWDC changes?

**Comment by LizCira** [comment]
@ohwutup I fixed my commit history and also fixed the style violations (IBuild is my work account) 
Let me know what else you need.

**Comment by sfransen** [comment]
@LizCira What I mean Is that is Get Error when I try and use this code after we had created more devices in License and more AppID an Provision profiles. Changing to 10 solved the problem I figure at one point I going to have to change to number even lower. we have 5000 devices in License and almost 1000 Profiles now and it take up a lot of memory just going thru the xxxx.all call.

File: Gemfile [hunk 1]
```diff
@@ -2,4 +2,4 @@ source "https://rubygems.org"
 
 gemspec path: "fastlane"
 
-gem "danger", "~> 0.8.1"
+gem "danger", "~> 0.8.1"
```
Comment (KrauseFx) [inline Gemfile]: This change shouldn't be here

File: fastlane_core/lib/fastlane_core/provisioning_profile.rb [hunk 2]
```diff
@@ -23,8 +23,8 @@ class << self
       #   "Version"=>1}
       def parse(path)
         require 'plist'
-
         plist = Plist.parse_xml(`security cms -D -i "#{path}"`)
+
```
Comment (KrauseFx) [inline fastlane_core/lib/fastlane_core/provisioning_profile.rb]: This change shouldn't be here
 Add a line note
View

File: sigh/lib/sigh/download_all.rb [hunk 3]
```diff
@@ -7,7 +7,7 @@ def download_all
       Spaceship.select_team
       UI.message "Successfully logged in"
 
-      Spaceship.provisioning_profile.all.each do |profile|
+      Spaceship.provisioning_profile.all(false).each do |profile|
```
Comment (KrauseFx) [inline sigh/lib/sigh/download_all.rb]: It's unclear what this parameter does. Should be a named parameter

Comment (KrauseFx) [inline sigh/lib/sigh/download_all.rb]: Should this set the `include_invalid_profiles` value? If so, why do we set it

File: sigh/lib/sigh/download_all.rb [hunk 4]
```diff
@@ -19,7 +19,12 @@ def download_all
 
     def download_profile(profile)
       FileUtils.mkdir_p(Sigh.config[:output_path])
-      profile_name = "#{profile.class.pretty_type}_#{profile.app.bundle_id}.mobileprovision" # default name
+
+      if Sigh.config[:use_apple_developer_portal_profile_names]
```
Comment (KrauseFx) [inline sigh/lib/sigh/download_all.rb]: This change has nothing to do with the pagination support, right? If so, can you move that out into a separate PR?

File: sigh/lib/sigh/options.rb [hunk 5]
```diff
@@ -96,6 +96,12 @@ def self.available_options
                                      is_string: false,
                                      default_value: false,
                                      short_option: "-w"),
+        FastlaneCore::ConfigItem.new(key: :use_apple_developer_portal_profile_names,
```
Comment (KrauseFx) [inline sigh/lib/sigh/options.rb]: Same as https://github.com/fastlane/fastlane/pull/4670/files#r67810814

File: spaceship/lib/spaceship/client.rb [hunk 6]
```diff
@@ -177,6 +177,7 @@ def page_size
     def paging
       page = 0
       results = []
+
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/client.rb]: This change shouldn't be here

File: spaceship/lib/spaceship/portal/portal_client.rb [hunk 7]
```diff
@@ -319,17 +319,24 @@ def revoke_certificate!(certificate_id, type, mac: false)
...
-    def provisioning_profiles(mac: false)
-      req = request(:post) do |r|
-        r.url "https://developerservices2.apple.com/services/#{PROTOCOL_VERSION}/#{platform_slug(mac)}/listProvisioningProfiles.action"
-        r.params = {
-          teamId: team_id,
-          includeInactiveProfiles: true,
-          onlyCountLists: true
-        }
+    def provisioning_profiles(include_invalid_profiles = true, mac: false)
+      @page_size = 10
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/portal/portal_client.rb]: Why is this a `@` variable?

File: spaceship/lib/spaceship/portal/portal_client.rb [hunk 8]
```diff
@@ -319,17 +319,24 @@ def revoke_certificate!(certificate_id, type, mac: false)
...
+      paging do |page_number|
+        puts "Pagination in progress, page number: '#{page_number}'..."
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/portal/portal_client.rb]: We don't log things in spaceship currently

File: spaceship/lib/spaceship/portal/provisioning_profile.rb [hunk 9]
```diff
@@ -242,8 +242,8 @@ def create!(name: nil, bundle_id: nil, certificate: nil, devices: [], mac: false
         # @return (Array) Returns all profiles registered for this account
         #  If you're calling this from a subclass (like AdHoc), this will
         #  only return the profiles that are of this type
-        def all(mac: false)
-          profiles = client.provisioning_profiles(mac: mac).map do |profile|
+        def all(include_invalid_profiles = true, mac: false)
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/portal/provisioning_profile.rb]: This should use named parameters

**Comment by LizCira** [comment]
@KrauseFx @ohwutup what's your preferred workflow for correcting PRs? I didn't see it in docs. Should I make corrections and push to the same branch or use the github editor? I'll try to take care of this over the weekend!

File: sigh/lib/sigh/download_all.rb (hunk 3 above)
Comment (LizCira) [inline sigh/lib/sigh/download_all.rb]: @KrauseFx I intended the default value to be "true", but the value for the download_all action to be false, since it speeds up download time a bit to skip invalid profiles. The HoundCI review didn't like it when I had the variable named, so I changed it to just the value. I can remove it entirely if that simplifies things.

**Comment by ohwutup** [comment]
Hey @LizCira! If you could just make the changes and push them up to this same branch that would be amazing! 🚀

**Comment by LizCira** [comment]
Hi @ohwutup are you waiting for anything from me for this?

**Comment by ohwutup** [comment]
@LizCira Yeah, I think last I knew, you were going to make those changes and push them up to this same branch right? Could you also rebase on master?

**Comment by sfransen** [comment]
is this ever going to be pushed to Master?

**Comment by ohwutup** [comment]
Hey @LizCira! Just wanted to check in here to see if you were still interested in getting this PR going!

**Comment by KrauseFx** [comment]
Hey everyone, so Apple has changed the way provisioning profiles can be fetched, so I'll close this PR for now. Feel free to submit a new one if changes are necessary 👍
